packages/plugins/Filter/src/utils/updateFeatureVisibility.ts

```diff
@@ -0,0 +1,148 @@
+  pattern: DatePattern
+): Date => {
+  const yearIndices = []
+  const monthIndices = []
+  const dayIndices = []
+  const indiceLookup = {
+    Y: yearIndices,
+    M: monthIndices,
+    D: dayIndices,
+  }
+  ;[...pattern].forEach((letter, index) => indiceLookup[letter]?.push?.(index))
+  function getFromPropertyValue(index) {
+    return propertyValue[index]
+  }
+  return new Date(
+    Number(yearIndices.map(getFromPropertyValue).join('')),
+    Number(monthIndices.map(getFromPropertyValue).join('')) - 1,
+    Number(dayIndices.map(getFromPropertyValue).join(''))
+  )
+}
+
+const doesFeaturePassTimeFilter = (
+  layerId: LayerId,
+  time: FilterState['time'],
+  timeOptions: TimeOption[],
+  feature: Feature
+): boolean => {
+  // radioId=0 means 'no restriction'
+  if (!time || time[layerId].radioId === 0) {
+    return true
+  }
+  const { targetProperty, radioId, pattern } = time[layerId]
+  const selectedTimeFilter = timeOptions[radioId - 1]
+  const propertyValue = feature.get(targetProperty)
+  const featureDate = getDateFromValue(propertyValue, pattern)
+  const { type, amount, unit } = selectedTimeFilter
+
+  // only unit 'days' currently supported
+  const unitMilliseconds = {
+    days: 24 * 60 * 60 * 1000,
+  }[unit]
+  const limits: Date[] = []
+
+  if (type === 'freeSelection') {
+    const chosenLimits = time[layerId].freeSelection
+    if (chosenLimits.length === 0) {
+      // no limits selected? feature passes automatically
+      return true
+    }
+    ;[limits[0], limits[1]] = getFreeSelectionLimits(chosenLimits)
+  } else {
+    limits[type === 'last' ? 0 : 1] = new Date(
+      // @ts-expect-error | amount is number in last/next case
+      Date.now() - amount * unitMilliseconds
+    )
+    limits[type === 'last' ? 1 : 0] = new Date(Date.now())
+  }
+  limits[0].setHours(0, 0, 0, 0)
+  limits[1].setHours(24, 0, 0, 0)
+
+  return limits[0] <= featureDate && featureDate <= limits[1]
+}
+
+const doesFeaturePassFilter = (
+  feature: Feature,
+  { category, time }: FilterState,
+  categories: FilterConfiguration['layers'][string]['categories'],
+  layerId: LayerId,
+  timeOptions: TimeOption[]
+): boolean => {
+  if (
+    category &&
+    !doesFeaturePassCategoryFilter(categories, category, feature, layerId)
+  ) {
+    return false
+  }
+
+  return Boolean(
+    time && doesFeaturePassTimeFilter(layerId, time, timeOptions, feature)
+  )
+}
+
+export const updateFeatureVisibility = ({
+  map,
+  layerId,
+  state,
+  categories,
+  timeOptions,
+}: {
+  map: Map
+  layerId: LayerId
+  state: FilterState
+  categories: FilterConfiguration['layers'][string]['categories']
+  timeOptions: TimeOption[]
+}) => {
+  const layer = map
+    .getLayers()
+    .getArray()
+    .find((layer) => layer.get('id') === layerId)
+  if (!layer) {
+    throw new Error(
+      `Layer ${layerId} undefined in Filter.utils.updateFeatureVisibility.`
+    )
+  }
+  layer
+    // @ts-expect-error | only layers with getSource allowed
```

Comment:
In X this is solved by typing the `layer` with `as VectorLayer<VectorSource>`. Typing it here intersected with `undefined` would probably be correct.

Also, one would have to check for `null` on `getSource` as well.